File: docs/snippets/all/archetypes/image_formats.rs
use rerun::external::ndarray;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rec = rerun::RecordingStreamBuilder::new("rerun_example_image_formats").spawn()?;

    // Simple gradient image
    let image = ndarray::Array3::from_shape_fn((256, 256, 3), |(y, x, c)| match c {
        0 => x as u8,
        1 => (x + y).min(255) as u8,
        2 => y as u8,
        _ => unreachable!(),
    });

    // RGB image
    rec.log(
        "image_rgb",
        &rerun::Image::from_color_model_and_tensor(rerun::ColorModel::RGB, image.clone())?,
    )?;

    // Green channel only (Luminance)
    rec.log(
        "image_green_only",
        &rerun::Image::from_color_model_and_tensor(
            rerun::ColorModel::L,
            image.slice(ndarray::s![.., .., 1]).to_owned(),
        )?,
    )?;

    // BGR image
    rec.log(
        "image_bgr",
        &rerun::Image::from_color_model_and_tensor(
            rerun::ColorModel::BGR,
            image.slice(ndarray::s![.., .., ..;-1]).to_owned(),
        )?,
    )?;

    // New image with Separate Y/U/V planes with 4:2:2 chroma downsampling
    let mut yuv_bytes = Vec::with_capacity(256 * 256 + 128 * 256 * 2);
    yuv_bytes.extend(std::iter::repeat(128).take(256 * 256)); // Fixed value for Y.
    yuv_bytes.extend((0..256).flat_map(|_y| (0..128).map(|x| x * 2))); // Gradient for U.
    yuv_bytes.extend((0..256).flat_map(|y| std::iter::repeat(y as u8).take(128))); // Gradient for V.
    rec.log(
        "image_yuv422",
        &rerun::Image::from_pixel_format(
            [256, 256],
            rerun::PixelFormat::Y_U_V16_FullRange,
            yuv_bytes,
        ),
    )?;

    Ok(())
}

File: docs/snippets/all/archetypes/image_formats.cpp
#include <algorithm>
#include <cstdint>
#include <vector>

#include <rerun.hpp>

int main(int argc, char* argv[]) {
    const auto rec = rerun::RecordingStream("rerun_example_image_formats");
    rec.spawn().exit_on_failure();

    // Simple gradient image
    std::vector<uint8_t> image(256 * 256 * 3);
    for (size_t y = 0; y < 256; ++y) {
        for (size_t x = 0; x < 256; ++x) {
            image[(y * 256 + x) * 3 + 0] = static_cast<uint8_t>(x);
            image[(y * 256 + x) * 3 + 1] = static_cast<uint8_t>(std::min<size_t>(255, x + y));
            image[(y * 256 + x) * 3 + 2] = static_cast<uint8_t>(y);
        }
    }

    // RGB image
    rec.log("image_rgb", rerun::Image::from_rgb24(image, {256, 256}));

    // Green channel only (Luminance)
    std::vector<uint8_t> green_channel(256 * 256);
    for (size_t i = 0; i < 256 * 256; ++i) {
        green_channel[i] = image[i * 3 + 1];
    }
    rec.log(
        "image_green_only",
        rerun::Image(rerun::borrow(green_channel), {256, 256}, rerun::ColorModel::L)
    );

    // BGR image
    std::vector<uint8_t> bgr_image(256 * 256 * 3);
    for (size_t i = 0; i < 256 * 256; ++i) {
        bgr_image[i * 3 + 0] = image[i * 3 + 2];
        bgr_image[i * 3 + 1] = image[i * 3 + 1];
        bgr_image[i * 3 + 2] = image[i * 3 + 0];
    }
    rec.log(
        "image_bgr",
        rerun::Image(rerun::borrow(bgr_image), {256, 256}, rerun::ColorModel::BGR)
    );

    // New image with Separate Y/U/V planes with 4:2:2 chroma downsampling
    std::vector<uint8_t> yuv_bytes(256 * 256 + 128 * 256 * 2);
    std::fill_n(yuv_bytes.begin(), 256 * 256, static_cast<uint8_t>(128)); // Fixed value for Y
    size_t u_plane_offset = 256 * 256;
    size_t v_plane_offset = u_plane_offset + 128 * 256;
    for (size_t y = 0; y < 256; ++y) {
        for (size_t x = 0; x < 128; ++x) {
            auto coord = y * 128 + x;
            yuv_bytes[u_plane_offset + coord] = static_cast<uint8_t>(x * 2); // Gradient for U
            yuv_bytes[v_plane_offset + coord] = static_cast<uint8_t>(y);     // Gradient for V
        }
    }
    rec.log(
        "image_yuv422",
        rerun::Image(rerun::borrow(yuv_bytes), {256, 256}, rerun::PixelFormat::Y_U_V16_FullRange)
    );

    return 0;
}
